dali/c_api/c_api.cc

```diff
@@ -64,47 +65,155 @@ void daliOutputRelease(daliPipelineHandle* pipe_handle) {
+}
+
+template <typename T>
+static size_t daliNumTensorsHelper(dali::DeviceWorkspace* ws, int n) {
+  return ws->Output<T>(n).ntensor();
+}
+
+size_t daliNumTensors(daliPipelineHandle* pipe_handle, int n) {
+  dali::DeviceWorkspace* ws = reinterpret_cast<dali::DeviceWorkspace*>(pipe_handle->ws);
+  size_t num_tensors = 0;
+  if (ws->OutputIsType<dali::CPUBackend>(n)) {
+    num_tensors = daliNumTensorsHelper<dali::CPUBackend>(ws, n);
+  } else {
+    num_tensors = daliNumTensorsHelper<dali::GPUBackend>(ws, n);
+  }
+  return num_tensors;
+}
+
+template <typename T>
+static size_t daliNumElementsHelper(dali::DeviceWorkspace* ws, int n) {
+  return ws->Output<T>(n).GetElmsNumber();
+}
+
+size_t daliNumElements(daliPipelineHandle* pipe_handle, int n) {
+  dali::DeviceWorkspace* ws = reinterpret_cast<dali::DeviceWorkspace*>(pipe_handle->ws);
+  size_t num_elms = 0;
+  if (ws->OutputIsType<dali::CPUBackend>(n)) {
+    num_elms = daliNumElementsHelper<dali::CPUBackend>(ws, n);
+  } else {
+    num_elms = daliNumElementsHelper<dali::GPUBackend>(ws, n);
+  }
+  return num_elms;
+}
+
+template <typename T>
+static size_t daliTensorSizeHelper(dali::DeviceWorkspace* ws, int n) {
+  return ws->Output<T>(n).nbytes();
+}
+
+size_t daliTensorSize(daliPipelineHandle* pipe_handle, int n) {
+  dali::DeviceWorkspace* ws = reinterpret_cast<dali::DeviceWorkspace*>(pipe_handle->ws);
+  size_t size = 0;
+  if (ws->OutputIsType<dali::CPUBackend>(n)) {
+    size = daliTensorSizeHelper<dali::CPUBackend>(ws, n);
+  } else {
+    size = daliTensorSizeHelper<dali::GPUBackend>(ws, n);
+  }
+  return size;
+}
+
+template <typename T>
+static size_t daliMaxDimTensorsHelper(dali::DeviceWorkspace* ws, int n) {
+  const auto &out_tensor_list = ws->Output<T>(n);
+  size_t tensors_num = out_tensor_list.ntensor();
+  size_t max_num_dim = 0;
+  for (size_t i = 0; i < tensors_num; ++i) {
+    size_t num_dim = out_tensor_list.tensor_shape(i).size();
+    // squeze last dimension
+    if (out_tensor_list.tensor_shape(i).back() == 1) {
+      --num_dim;
+    }
+    max_num_dim = std::max(max_num_dim, num_dim);
+  }
+  return max_num_dim;
+}
+
+size_t daliMaxDimTensors(daliPipelineHandle* pipe_handle, int n) {
+  dali::DeviceWorkspace* ws = reinterpret_cast<dali::DeviceWorkspace*>(pipe_handle->ws);
+  size_t max_num_dim = 0;
+  if (ws->OutputIsType<dali::CPUBackend>(n)) {
+    max_num_dim = daliMaxDimTensorsHelper<dali::CPUBackend>(ws, n);
+  } else {
+    max_num_dim = daliMaxDimTensorsHelper<dali::GPUBackend>(ws, n);
+  }
+  return max_num_dim;
+}
+
 unsigned daliGetNumOutput(daliPipelineHandle* pipe_handle) {
   dali::DeviceWorkspace* ws = reinterpret_cast<dali::DeviceWorkspace*>(pipe_handle->ws);
   return ws->NumOutput();
 }
 
+template <typename T>
+static void daliCopyTensorListNToHelper(dali::DeviceWorkspace* ws, void* dst, int n,
+                                    device_type_t dst_type) {
+  dali::CopyToExternalTensor(&(ws->Output<T>(n)), dst, (dali::device_type_t)dst_type);
```

Comment:
isn't `dst_type` already a `device_type_t` ?

Reply:
`device_type_t` is where data is stored, `dst_type` is where we want to copy data to.